pkg/kubelet/cm/memorymanager/memory_manager.go

```diff
@@ -0,0 +1,419 @@
+func (m *manager) removeStaleState() {
+	// Only once all sources are ready do we attempt to remove any stale state.
+	// This ensures that the call to `m.activePods()` below will succeed with
+	// the actual active pods list.
+	if !m.sourcesReady.AllReady() {
+		return
+	}
+
+	// We grab the lock to ensure that no new containers will grab memory block while
+	// executing the code below. Without this lock, its possible that we end up
+	// removing state that is newly added by an asynchronous call to
+	// AddContainer() during the execution of this code.
+	m.Lock()
+	defer m.Unlock()
+
+	// Get the list of active pods.
+	activePods := m.activePods()
+
+	// Build a list of (podUID, containerName) pairs for all containers in all active Pods.
+	activeContainers := make(map[string]map[string]struct{})
+	for _, pod := range activePods {
+		activeContainers[string(pod.UID)] = make(map[string]struct{})
+		for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+			activeContainers[string(pod.UID)][container.Name] = struct{}{}
+		}
+	}
+
+	// Loop through the MemoryManager state. Remove any state for containers not
+	// in the `activeContainers` list built above.
+	assignments := m.state.GetMemoryAssignments()
+	for podUID := range assignments {
+		for containerName := range assignments[podUID] {
+			if _, ok := activeContainers[podUID][containerName]; !ok {
+				klog.Infof("[memorymanager] removeStaleState: removing (pod %s, container: %s)", podUID, containerName)
+				err := m.policyRemoveContainerByRef(podUID, containerName)
+				if err != nil {
+					klog.Errorf("[memorymanager] removeStaleState: failed to remove (pod %s, container %s), error: %v)", podUID, containerName, err)
+				}
+			}
+		}
+	}
+}
+
+func (m *manager) policyRemoveContainerByRef(podUID string, containerName string) error {
+	err := m.policy.RemoveContainer(m.state, podUID, containerName)
+	if err == nil {
+		m.containerMap.RemoveByContainerRef(podUID, containerName)
+	}
+
+	return err
+}
+
+func getTotalMemoryTypeReserved(machineInfo *cadvisorapi.MachineInfo, reservedMemory map[int]map[v1.ResourceName]resource.Quantity) map[v1.ResourceName]resource.Quantity {
+	totalMemoryType := map[v1.ResourceName]resource.Quantity{}
+
+	numaNodes := map[int]bool{}
+	for _, numaNode := range machineInfo.Topology {
+		numaNodes[numaNode.Id] = true
+	}
+
+	for nodeID, node := range reservedMemory {
+		if !numaNodes[nodeID] {
+			klog.Warningf("The NUMA node %d specified under --reserved- memory does not exist on the machine", nodeID)
+			continue
+		}
+
+		for memType, memVal := range node {
+			if totalMem, exists := totalMemoryType[memType]; exists {
+				memVal.Add(totalMem)
+			}
+			totalMemoryType[memType] = memVal
+		}
+	}
+
+	return totalMemoryType
+}
+
+func validateReservedMemory(machineInfo *cadvisorapi.MachineInfo, nodeAllocatableReservation v1.ResourceList, reservedMemory map[int]map[v1.ResourceName]resource.Quantity) error {
```

Comment:
We don't have to address this now, but I'm wondering if this validation should actually happen up near the API layer rather than inside this component.

Reply:
I agree with you, for now, I just followed the pattern regarding `reservedCpus` under the CPU manager. But IIUC moving it near the API layer can work only when the KubeletConfiguration used and not CLI arguments.